Zend/Optimizer/zend_optimizer.c

```diff
@@ -891,16 +891,20 @@ zend_function *zend_optimizer_get_called_func(
 					&op_array->scope->function_table, method_name);
 				if (fbc) {
 					bool is_private = (fbc->common.fn_flags & ZEND_ACC_PRIVATE) != 0;
-					bool is_final = (fbc->common.fn_flags & ZEND_ACC_FINAL) != 0;
-					bool same_scope = fbc->common.scope == op_array->scope;
 					if (is_private) {
 						/* Only use private method if in the same scope. We can't even use it
 						 * as a prototype, as it may be overridden with changed signature. */
+						bool same_scope = fbc->common.scope == op_array->scope;
 						return same_scope ? fbc : NULL;
 					}
-					/* If the method is non-final, it may be overridden,
-					 * but only with a compatible method signature. */
-					*is_prototype = !is_final;
+					/* We know it is a prototype if it can be overridden with a compatible signature.
+					 * A method cannot be overridden if the class or method is final, or
+					 * if it's an anonymous class. */
+					bool can_be_overridden = (fbc->common.fn_flags & ZEND_ACC_FINAL) == 0 &&
+											 (fbc->common.scope->ce_flags & (ZEND_ACC_FINAL | ZEND_ACC_ANON_CLASS)) == 0;
```

Comment:
I wonder if there are other pieces of code that rely on that information. For example, wouldn't your suggestion influence the information you get from reflection? e.g. it might be unexpected that the user asks "is this method final" and gets yes even though they only used final on the class definition and not on the method.
But idk, I'm just speculating.

Reply:
Yeah, basically I suggested the same in the other PR :) But I guess then some other special cases should be added in reflection (?) or when the code is dumped.. so maybe doesn't worth it

Reply:
It would indeed influence reflection. But IMO a `final` flag for anonymous classes would make sense, as it fulfills the meaning of `final` for normal classes.

Reply:
I just learned that anonymous classes are not final, an anonymous class can even extend another anonymous class. (See: Zend/tests/anon/011.phpt). So performing that change breaks that test, and I think that test is legal behaviour that should be allowed. I will therefore not make this change.
I will also correct the check I put in the current code.

Reply:
Pushed the correction, and comment update.

Reply:
> an anonymous class can even extend another anonymous class. (See: Zend/tests/anon/011.phpt)

https://media.giphy.com/media/8OlT82jKm6Ugg/giphy.gif

Thank you :slightly_smiling_face: